text/0001-blinking-cursors-and-periodic-animation.md

```diff
@@ -0,0 +1,182 @@
+
+Modifying the behavior of the wait state is surprisingly trivial, thanks to the underlying use of `winit`. Instead of setting the event loops' `control_flow` to `Wait`, we just set it to `WaitUntil( <time of soonest update> )`. This changes the behavior to wait until the next event, or the provided time (whichever is sooner).
+
+As the event loop only updates/draws in response to events/user-interaction, we emit a new synthetic event `AnimationTick` to drive the update cycle.
+
+### 2: Widget trait changes
+
+An update is needed to the [Widget](https://docs.rs/iced_native/0.2.2/iced_native/widget/trait.Widget.html) trait to communicate the animation requirements of a widget - in our case, that the `TextInput` widget needs to be re-drawn in 500ms.
+
+We propose adding a new method to the trait, with a default implementation that just indicate no animation is taking place:
+
+```rust
+  fn next_animation(&self) -> AnimationState {
+    AnimationState::NotAnimating
+  }
+```
+
+Widgets that need to animate (such as our `TextInput`) can implement this method to indicate an animation is required:
+
+```rust
+  fn next_animation(&self) -> AnimationState {
+    AnimationState::AnimateIn(std::time::Instant::now().checked_add(Duration::from_millis(500)))
+  }
+```
+
+Remaining consistent with the stateless and intuitive feel of the widget trait, `next_animation` is called as part of the update loop, to get the latest set of animation requirements.
+
+### 3: `TextInput` changes
+
+The widget needs to keep track of the last time a user interacted with the input. We can do this by adding a new field to an internal state type `Cursor`:
+
+```rust
+pub struct Cursor {
+    state: State,
+    updated_at: Instant, // new field
+}
+```
+
+And setting its value when the cursor state is updated:
+
+```rust
+pub(crate) fn move_to(&mut self, position: usize) {
+    self.updated_at = Instant::now(); // add this line
+    self.state = State::Index(position);
+}
+```
+
+As a corner case, we also need to detect when the input is clicked and gains focus, which we can do by setting `updated_at` during `on_event` when we detect that condition.
+
+### The `AnimationState` type
+
+Widgets need to symbolize their animation requirements to the runtime. We propose creating a new enum type to represent this:
+
+```rust
+pub enum AnimationState {
+    NotAnimating, // The widget does not need to animate itself, and will only change in response to events.
+    AnimateIn(std::time::Instant), // The widget needs to animate itself no sooner than the provided moment.
+}
+```
+
+A new type seems ideal for the following reasons:
+
+1. It can be easily extended to support future animation use-cases (for instance, continuous/every-frame animations)
```

Comment:
I believe it could be used for frame invalidation / redraw requests. In fact, I think it may be a great idea to rename the enum to `RedrawRequest` or similar:

```rust
pub enum RedrawRequest {
    None,
    Now,
    At(Instant),
}
```

`iced` currently redraws on every iteration of the event loop. The work here could allow us to instead redraw (everything) only when a widget needs it. Even though this is probably out of scope for this first iteration, we can try to start opening that door.

Reply:
Yeah that makes sense. For clarity, you mean (by 'redraw' in this context) literally running the graphics pump / swapchain to emit a new frame (as opposed to computing the view tree via invoking `view()` or some other meaning of 'redraw')?